### DISCUSSION TITLE
[javascript deep dive] 4장. 변수

### DISCUSSION THREAD

**Discussion by rladpwl0512** [woowacourse-study/2022-javascript-bank#2] [General]
[javascript deep dive] 4장. 변수

**Comment by jswith** [comment]
var, let, const의 차이에 대해 설명해보세요

**Reply by jswith** [reply]
- var : 함수 스코프, 재선언, 재할당 가능, 호이스팅 -> 할당 전 undefined 반환
- let, const: 블록 스코프, 호이스팅 안되는 것 처럼 보임, TDZ -> 할당 전 에러
- let은 재할당 가능
- const는 재할당 불가

**Comment by jin7969** [comment]
변수를 선언하고 재할당 할 경우 같은 메모리에 값을 삭제하고 새로운 값을 저장하면 되지,
왜 새로운 메모리에 저장하나요?
![스크린샷 2022-09-09 오후 4 16 12](https://user-images.githubusercontent.com/82227098/189293340-ea9a6691-16fd-4eec-8299-42d017a8567f.png)

**Reply by jin7969** [reply]
변수 a,b가 같은 값을 바라볼때,
a를 변경하면 a만 변경되고 b는 변경되지 않아야 하기 때문에

**Reply by rladpwl0512** [reply]
const a = 3; 
const b = 3; 
일 때, a, b 각각이 3을 메모리셀에 새롭게 할당하는가? 아니면 하나의 3을 a, b 둘다 가리키고있는가? 

-> 자바스크립트 엔진 동작은 정확히 모른다 (deep dive)
하지만 코어 javascript, mdn에서는 같은 곳을 가리킨다고 한다.

**Comment by rladpwl0512** [comment]
변수란 무엇이고, 왜 필요한가요?

**Reply by rladpwl0512** [reply]
- 변수 내부에 상수가 있다. 
- 변수 (변수에는 변하는 값=변수, 변하지 않는 값이 있다.=상수) 
- 상수: 변하지 않는 변수 

- 이름을 붙히지 않으면, 값이 어디에 저장되어있는지 알기 힘들다. 변수와 같은 식별자를 붙힘으로써 원하는 값에 쉽게 접근이 가능하다. 
- 변수를 통해서만 값에 접근할 수 있나요? -> 메모리 주소를 통해서도 접근할 수 있지만, 메모리 주소를 통해 값에 직접 접근하는 것은 실수로 운영체제가 사용하는 중요한 값을 변경할 수 있고, 이는 시스템을 멈추게 하는 치명적 오류를 발생시킬 수 있다.

**Comment by rladpwl0512** [comment]
호이스팅이란 무엇인가요? (var / let, const는 호이스팅이 일어나나요?)

**Reply by rladpwl0512** [reply]
- 호이스팅이란? 
선언부가 코드 가장 위로 끌어올려지는 현상 

- 호이스팅을 왜 하는가? 
변수가 무엇이 있는지 미리 확인 및 필요한 메모리 공간을 미리 할당할 수 있다. (메모리 공간 효율적으로 사용가능) 

- 함수 선언식, 함수 표현식 
함수표현식을 var -> undefined (함수표현식 중, var는 호이스팅이 된다!) 
함수표현식을 const -> reference 에러
